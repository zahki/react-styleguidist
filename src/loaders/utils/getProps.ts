import path from 'path';
import fs from 'fs';
import { TagProps, TagParamObject, DocumentationObject, utils, TagObject } from 'react-docgen';
import _ from 'lodash';
import doctrine, { Annotation } from 'doctrine';
import createLogger from 'glogg';
import highlightCodeInMarkdown from './highlightCodeInMarkdown';
import removeDoclets from './removeDoclets';
import requireIt from './requireIt';
import getNameFromFilePath from './getNameFromFilePath';
import * as Rsg from '../../typings';

const logger = createLogger('rsg');

const examplesLoader = path.resolve(__dirname, '../examples-loader.js');

const JS_DOC_METHOD_PARAM_TAG_SYNONYMS: (keyof TagProps)[] = ['param', 'arg', 'argument'];
const JS_DOC_METHOD_RETURN_TAG_SYNONYMS: (keyof TagProps)[] = ['return', 'returns'];
const JS_DOC_ALL_SYNONYMS: (keyof TagProps)[] = [
	...JS_DOC_METHOD_PARAM_TAG_SYNONYMS,
	...JS_DOC_METHOD_RETURN_TAG_SYNONYMS,
];

// HACK: We have to make sure that doclets is a proper object with correct prototype to
// work around an issue in react-docgen that breaks the build if a component has JSDoc tags
// like @see in its description, see https://github.com/reactjs/react-docgen/issues/155
// and https://github.com/styleguidist/react-styleguidist/issues/298
const getDocletsObject = (str?: string) => ({ ...utils.docblock.getDoclets(str) });

const getDoctrineTags = (documentation: Annotation) => {
	return _.groupBy(documentation.tags, 'title');
};

const doesExternalExampleFileExist = (componentPath: string, exampleFile: string) => {
	const exampleFilepath = path.resolve(path.dirname(componentPath), exampleFile);
	const doesFileExist = fs.existsSync(exampleFilepath);

	if (!doesFileExist) {
		logger.warn(`An example file ${exampleFile} defined in ${componentPath} component not found.`);
	}
	return doesFileExist;
};

const getMergedTag = (tags: TagProps, names: (keyof TagProps)[]): TagObject[] => {
	return names.reduce((params: TagObject[], name) => [...params, ...(tags[name] || [])], []);
};

/**
 * 1. Remove non-public methods.
 * 2. Extract doclets.
 * 3. Highlight code in descriptions.
 * 4. Extract @example doclet (load linked file with examples-loader).
 *
 * @param {object} doc
 * @param {string} filepath
 * @returns {object}
 */
export default function getProps(doc: DocumentationObject, filepath?: string): Rsg.TempPropsObject {
	const outDocs: Rsg.TempPropsObject = { doclets: {}, displayName: '', ...doc, methods: undefined };

	// Keep only public methods
	outDocs.methods = (doc.methods || []).filter(method => {
		const doclets = method.docblock && utils.docblock.getDoclets(method.docblock);
		return doclets && doclets.public;
	}) as Rsg.MethodWithDocblock[];

	// Parse the docblock of the remaining methods with doctrine to retrieve
	// the JSDoc tags
	// if a method is visible it must have a docblock
	outDocs.methods = outDocs.methods.map(method => {
		const allTags = getDoctrineTags(
			doctrine.parse(method.docblock, { sloppy: true, unwrap: true })
		);

		// Merge with react-docgen information about arguments and return value
		// with information from JSDoc

		const paramTags = getMergedTag(
			allTags as TagProps,
			JS_DOC_METHOD_PARAM_TAG_SYNONYMS
		) as TagParamObject[];
		const params =
			method.params &&
			method.params.map(param => ({
				...param,
				...paramTags.find(tagParam => tagParam.name === param.name),
			}));

		if (params) {
			method.params = params;
		}

		const returnTags = getMergedTag(
			allTags as TagProps,
			JS_DOC_METHOD_RETURN_TAG_SYNONYMS
		) as TagParamObject[];
		const returns = method.returns
			? {
					...method.returns,
					type: {
						type: 'NameExpression',
						...method.returns.type,
					},
			  }
			: returnTags[0];

		if (returns) {
			method.returns = returns;
		}

		// Remove tag synonyms
		method.tags = _.omit(allTags, JS_DOC_ALL_SYNONYMS);

		return method;
	});

	if (doc.description) {
		// Read doclets from the description and remove them
		outDocs.doclets = getDocletsObject(doc.description);

		const documentation = doctrine.parse(doc.description);
		outDocs.tags = getDoctrineTags(documentation) as TagProps;

		outDocs.description = highlightCodeInMarkdown(removeDoclets(doc.description));

		let exampleFileExists = false;
		let exampleFile = outDocs.doclets.example;

		// doc.doclets.example might be a boolean or undefined
		if (typeof outDocs.doclets.example === 'string' && filepath) {
			exampleFile = outDocs.doclets.example.trim();
			exampleFileExists = doesExternalExampleFileExist(filepath, exampleFile);
		}

		if (exampleFileExists) {
			outDocs.example = requireIt(`!!${examplesLoader}!${exampleFile}`);
			delete outDocs.doclets.example;
		}
	} else {
		outDocs.doclets = {};
	}

	if (doc.props) {
		// Read doclets of props
		Object.keys(doc.props).forEach(propName => {
			if (!doc.props) {
				return;
			}
			const prop = doc.props[propName];
			const doclets = getDocletsObject(prop.description);

			// When a prop is listed in defaultProps but not in props the prop.description is undefined
			const documentation = doctrine.parse(prop.description || '');

			// documentation.description is the description without tags
			prop.description = documentation.description;
			prop.tags = getDoctrineTags(documentation) as TagProps;

			// Remove ignored props
			if (doclets && doclets.ignore && outDocs.props) {
				delete outDocs.props[propName];
			} else if (outDocs.props) {
				outDocs.props[propName] = prop;
			}
		});
	}

	if (!doc.displayName && filepath) {
		// Guess the exported component's display name based on the file path
		outDocs.displayName = getNameFromFilePath(filepath);
	}

	if (outDocs.doclets && outDocs.doclets.visibleName) {
		outDocs.visibleName = outDocs.doclets.visibleName;

		// Custom tag is added both to doclets and tags
		// Removing from both locations
		delete outDocs.doclets.visibleName;
		if (outDocs.tags) {
			delete outDocs.tags.visibleName;
		}
	}

	return outDocs;
}
