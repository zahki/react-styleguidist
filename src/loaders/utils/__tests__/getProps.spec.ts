import path from 'path';
import getProps from '../getProps';

it('should return an object for props', () => {
	const result = getProps({
		displayName: 'Button',
		description: 'The only true button.',
		methods: [],
		props: {
			children: {
				name: 'children',
				type: {
					name: 'object',
				},
				required: true,
				description: 'Button label.',
			},
			color: {
				name: 'color',
				type: {
					name: 'string',
				},
				required: false,
				description: '',
			},
		},
	});

	expect(result).toMatchSnapshot();
});

it('should return an object for props without description', () => {
	const result = getProps({
		displayName: 'Button',
		props: {
			children: {
				name: 'children',
				type: {
					name: 'object',
				},
				required: true,
				description: 'Button label.',
			},
		},
	});

	expect(result).toMatchSnapshot();
});

it('should remove non-public methods', () => {
	const result = getProps(
		{
			displayName: 'Button',
			methods: [
				{
					docblock: `Public method.
@public`,
				},
				{
					docblock: `Private method.
@private`,
				},
				{
					docblock: 'Private method by default.',
				},
			] as any,
		},
		__filename
	);

	expect(result).toMatchSnapshot();
});

it('should get method info from docblock and merge it', () => {
	const result = getProps(
		{
			displayName: 'Button',
			methods: [
				{
					docblock: `
Baz method with foo param

@public
@returns {string} test
`,
				},
			] as any,
		},
		__filename
	);

	expect(result.methods).toMatchSnapshot();
});

it('should accept @return as a synonym of @returns', () => {
	const result = getProps(
		{
			displayName: 'Button',
			methods: [
				{
					docblock: `
Baz method with foo param

@public
@return {string} test
`,
				},
			] as any,
		},
		__filename
	);

	expect(result.methods).toMatchSnapshot();
});

it('should get method params info from docblock and merge it with passed method info', () => {
	const result = getProps(
		{
			displayName: 'Button',
			methods: [
				{
					docblock: `
Foo method with baz param

@public
@param {string} [baz=bar]
@arg {string} foo param described with @arg tag
@argument {string} test param described with @argument tag
@returns {string} test
`,
					params: [
						{
							name: 'baz',
						},
						{
							name: 'foo',
						},
						{
							name: 'test',
						},
					],
				},
			] as any,
		},
		__filename
	);

	expect(result.methods).toMatchSnapshot();
});

it('should return an object for props with doclets', () => {
	const result = getProps(
		{
			displayName: 'Button',
			description: `
The only true button.

@foo Foo
@bar Bar
`,
		},
		__filename
	);

	expect(result).toMatchSnapshot();
});

it('should return require statement for @example doclet', () => {
	const result = getProps(
		{
			displayName: 'Button',
			description: `
The only true button.

@example ../../../test/components/Placeholder/examples.md
`,
		},
		__filename
	);

	expect(result).toMatchSnapshot();
});

it('should return require statement for @example doclet only when the file exists', () => {
	const result = getProps(
		{
			displayName: 'Button',
			description: `
The only true button.

@example example.md
`,
		},
		__filename
	);

	expect(result).toMatchSnapshot();
});

it('should highlight code in description (fenced code block)', () => {
	const result = getProps({
		description: `
The only true button.

\`\`\`js
alert('Hello world');
\`\`\`
`,
	});

	expect(result).toMatchSnapshot();
});

it("should not crash when using doctrine to parse a default prop that isn't in the props list", () => {
	const result = getProps({
		description: 'The only true button.',
		methods: [],
		props: {
			crash: {
				description: undefined,
			},
		} as any,
	});

	expect(result).toMatchSnapshot();
});

it('should not crash when using doctrine to parse a return method that does not have type in it', () => {
	const result = getProps(
		{
			displayName: 'Button',
			methods: [
				{
					docblock: `
Public Method

@public
@returns {Boolean} return a Boolean Value
`,
					returns: {
						description: 'return a Boolean Value',
						type: { name: 'boolean' },
					},
				},
			] as any,
		},
		__filename
	);

	// @ts-ignore
	expect(result.methods[0].returns).toEqual(
		expect.objectContaining({
			description: 'return a Boolean Value',
			type: {
				name: 'boolean',
				type: 'NameExpression',
			},
		})
	);
});

it('should guess a displayName for components that react-docgen was not able to recognize', () => {
	const result = getProps(
		{
			methods: [],
			props: {},
		},
		path.join('an', 'absolute', 'path', 'to', 'YourComponent.js')
	);
	expect(result).toHaveProperty('displayName', 'YourComponent');
});

describe('with @visibleName tag present in the description', () => {
	const result = getProps({
		description: 'bar\n@visibleName foo',
	});
	it('should set visibleName property on the docs object', () => {
		expect(result).toHaveProperty('visibleName', 'foo');
	});

	it('should delete visibleName from doclets on the docs object', () => {
		expect(result.doclets).not.toHaveProperty('visibleName');
	});

	it('should delete visibleName from tags on the docs object', () => {
		expect(result.tags).not.toHaveProperty('visibleName');
	});
});
