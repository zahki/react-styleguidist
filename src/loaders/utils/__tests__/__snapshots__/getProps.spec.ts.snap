// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept @return as a synonym of @returns 1`] = `
Array [
  Object {
    "docblock": "
Baz method with foo param

@public
@return {string} test
",
    "returns": Object {
      "description": "test",
      "title": "return",
      "type": Object {
        "name": "string",
        "type": "NameExpression",
      },
    },
    "tags": Object {
      "public": Array [
        Object {
          "description": null,
          "title": "public",
          "type": null,
        },
      ],
    },
  },
]
`;

exports[`should get method info from docblock and merge it 1`] = `
Array [
  Object {
    "docblock": "
Baz method with foo param

@public
@returns {string} test
",
    "returns": Object {
      "description": "test",
      "title": "returns",
      "type": Object {
        "name": "string",
        "type": "NameExpression",
      },
    },
    "tags": Object {
      "public": Array [
        Object {
          "description": null,
          "title": "public",
          "type": null,
        },
      ],
    },
  },
]
`;

exports[`should get method params info from docblock and merge it with passed method info 1`] = `
Array [
  Object {
    "docblock": "
Foo method with baz param

@public
@param {string} [baz=bar]
@arg {string} foo param described with @arg tag
@argument {string} test param described with @argument tag
@returns {string} test
",
    "params": Array [
      Object {
        "default": "bar",
        "description": null,
        "name": "baz",
        "title": "param",
        "type": Object {
          "expression": Object {
            "name": "string",
            "type": "NameExpression",
          },
          "type": "OptionalType",
        },
      },
      Object {
        "description": "param described with @arg tag",
        "name": "foo",
        "title": "arg",
        "type": Object {
          "name": "string",
          "type": "NameExpression",
        },
      },
      Object {
        "description": "param described with @argument tag",
        "name": "test",
        "title": "argument",
        "type": Object {
          "name": "string",
          "type": "NameExpression",
        },
      },
    ],
    "returns": Object {
      "description": "test",
      "title": "returns",
      "type": Object {
        "name": "string",
        "type": "NameExpression",
      },
    },
    "tags": Object {
      "public": Array [
        Object {
          "description": null,
          "title": "public",
          "type": null,
        },
      ],
    },
  },
]
`;

exports[`should highlight code in description (fenced code block) 1`] = `
Object {
  "description": "The only true button.

\`\`\`js
<span class=\\"token function\\">alert</span><span class=\\"token punctuation\\">(</span><span class=\\"token string\\">'Hello world'</span><span class=\\"token punctuation\\">)</span><span class=\\"token punctuation\\">;</span>
\`\`\`
",
  "displayName": "",
  "doclets": Object {},
  "methods": Array [],
  "tags": Object {},
}
`;

exports[`should not crash when using doctrine to parse a default prop that isn't in the props list 1`] = `
Object {
  "description": "The only true button.
",
  "displayName": "",
  "doclets": Object {},
  "methods": Array [],
  "props": Object {
    "crash": Object {
      "description": "",
      "tags": Object {},
    },
  },
  "tags": Object {},
}
`;

exports[`should remove non-public methods 1`] = `
Object {
  "displayName": "Button",
  "doclets": Object {},
  "methods": Array [
    Object {
      "docblock": "Public method.
@public",
      "tags": Object {
        "public": Array [
          Object {
            "description": null,
            "title": "public",
          },
        ],
      },
    },
  ],
}
`;

exports[`should return an object for props 1`] = `
Object {
  "description": "The only true button.
",
  "displayName": "Button",
  "doclets": Object {},
  "methods": Array [],
  "props": Object {
    "children": Object {
      "description": "Button label.",
      "name": "children",
      "required": true,
      "tags": Object {},
      "type": Object {
        "name": "object",
      },
    },
    "color": Object {
      "description": "",
      "name": "color",
      "required": false,
      "tags": Object {},
      "type": Object {
        "name": "string",
      },
    },
  },
  "tags": Object {},
}
`;

exports[`should return an object for props with doclets 1`] = `
Object {
  "description": "The only true button.
",
  "displayName": "Button",
  "doclets": Object {
    "bar": "Bar
",
    "foo": "Foo",
  },
  "methods": Array [],
  "tags": Object {
    "bar": Array [
      Object {
        "description": "Bar",
        "title": "bar",
      },
    ],
    "foo": Array [
      Object {
        "description": "Foo",
        "title": "foo",
      },
    ],
  },
}
`;

exports[`should return an object for props without description 1`] = `
Object {
  "displayName": "Button",
  "doclets": Object {},
  "methods": Array [],
  "props": Object {
    "children": Object {
      "description": "Button label.",
      "name": "children",
      "required": true,
      "tags": Object {},
      "type": Object {
        "name": "object",
      },
    },
  },
}
`;

exports[`should return require statement for @example doclet 1`] = `
Object {
  "description": "The only true button.
",
  "displayName": "Button",
  "doclets": Object {
    "example": "../../../test/components/Placeholder/examples.md
",
  },
  "methods": Array [],
  "tags": Object {
    "example": Array [
      Object {
        "description": "../../../test/components/Placeholder/examples.md",
        "title": "example",
      },
    ],
  },
}
`;

exports[`should return require statement for @example doclet only when the file exists 1`] = `
Object {
  "description": "The only true button.
",
  "displayName": "Button",
  "doclets": Object {
    "example": "example.md
",
  },
  "methods": Array [],
  "tags": Object {
    "example": Array [
      Object {
        "description": "example.md",
        "title": "example",
      },
    ],
  },
}
`;
